### PR TITLE
Fixes  #71 Extend Schema with PKRatio

### LINKED ISSUE #71: Extend Schema with PKRatio
https://github.com/Open-Systems-Pharmacology/QualificationPlan/issues/19

### LINKED ISSUE #82: Extend schema with plot axes definitions
Axes will be defined globally (per plot type).

<details><summary>Global axes settings (<b>expand to see an example</b>)</summary>

```
    "AxesSettings": {
      "ComparisonTimeProfile": [
        {
          "Unit": "h",
          "Dimension": "Time",
          "Type": "X",
          "GridLines": false,
          "Scaling": "Linear"
        },
        {
          "Unit": "µg/ml",
          "Dimension": "Concentration (mass)",
          "Type": "Y",
          "GridLines": false,
          "Scaling": "Log"
        }
      ],
      "DDIRatioPlots_PredictedVsObserved": [
        {
          "Unit": "",
          "Dimension": "Dimensionless",
          "Type": "X",
          "GridLines": false,
          "Scaling": "Log"
        },
        {
          "Unit": "",
          "Dimension": "Dimensionless",
          "Type": "Y",
          "GridLines": false,
          "Scaling": "Log"
        }
      ],
      "DDIRatioPlots_ResidualsVsObserved": [
        {
          "Unit": "",
          "Dimension": "Dimensionless",
          "Type": "X",
          "GridLines": false,
          "Scaling": "Log"
        },
        {
          "Unit": "",
          "Dimension": "Dimensionless",
          "Type": "Y",
          "GridLines": false,
          "Scaling": "Linear"
        }
      ],
      "GOFMergedPlots_PredictedVsObserved": [
        {
          "Unit": "µg/ml",
          "Dimension": "Concentration (mass)",
          "Type": "X",
          "GridLines": false,
          "Scaling": "Log"
        },
        {
          "Unit": "µg/ml",
          "Dimension": "Concentration (mass)",
          "Type": "Y",
          "GridLines": false,
          "Scaling": "Log"
        }
      ],
      "GOFMergedPlots_ResidualsOverTime": [
        {
          "Unit": "h",
          "Dimension": "Time",
          "Type": "X",
          "GridLines": false,
          "Scaling": "Linear"
        },
        {
          "Unit": "",
          "Dimension": "Dimensionless",
          "Type": "Y",
          "GridLines": false,
          "Scaling": "Linear"
        }
      ],
      "PKRatioPlots": [
        {
          "Unit": "year(s)",
          "Dimension": "Age",
          "Type": "X",
          "GridLines": false,
          "Scaling": "Linear"
        },
        {
          "Unit": "",
          "Dimension": "Dimensionless",
          "Type": "Y",
          "GridLines": false,
          "Scaling": "Log"
        }
      ]
    }
```
</details>

Additionally each plot can define local axes settings which will override the global settings.
Following axes settings can be optionally defined per plot:
* GOFMergedPlots
  * AxesPredictedVsObserved
  * AxesResidualsOverTime
* ComparisonTimeProfilePlots
  * Axes
* DDIRatioPlots
  * AxesPredictedVsObserved
  * AxesResidualsVsObserved
* PKRatioPlots
  * Axes

### LINKED ISSUE #83: Extend advanced example with axes definitions

### LINKED ISSUE #84: Extend minimal example with axes definitions

### LINKED ISSUE #85: Update DDI Ratio observed data with new format and adjust advanced example

### PR DESCRIPTION
**Pull request by Yuri05**
Fixes #82 Extend schema with plot axes definitions
Fixes #83 Extend advanced example with axes definitions
Fixes #84 Extend minimal example with axes definitions
Fixes #85 Update DDI Ratio observed data with new format and adjust advanced example

### CODE REVIEW

File: examples/advanced_01/input/DDI_Ratio_CYP3A4.csv [hunk 1]
```diff
@@ -0,0 +1,11 @@
+ID,Study ID,Reference,Grouping,Substrate,Perpetrator,Route Substrate,Route Perpetrator,Compartment,AUCR Avg,AUCR AvgType,AUCR Var,AUCR VarType,CmaxR Avg,CmaxR AvgType,CmaxR Var,CmaxR VarType,t_placebo_0,t_placebo_end,t_treatment_0,t_treatment_end,t Unit,Comment,Dose,Dose free API,Dose Unit,Route,Times of Administration,Times Unit,Comment Regimen,Administered form,Formulation type,Infusion duration [min],Water volume ingested for drug intake [mL],Comment on administration/formulation,Fasted/Fed state,Duration of fasting before drug administration [h],Duration of fasting after drug administration [h],Comment on food intake
+50,Ahonen 1995,http://www.ncbi.nlm.nih.gov/pubmed/6138081,with Perpetrator (Itraconazole),Midazolam,Itraconazole,0,,Plasma,5.75,,,,2.56,,,,,,,,,,200,,mg,PO,"(S-0,T-24,R-15)",h,OD,,Sporanox Oral Solution,,200,hydroxypropyl-beta-cyclodextrin solution (HP-beta-CD) of 20 ml,,,,
+56,Backman 1998,https://www.ncbi.nlm.nih.gov/pubmed/9591931,Phase II (during Perpetrator (Rifampicin)),Midazolam,Rifampicin,0,,Plasma,7.97,,,,3.12,,,,,,,,,,,,,,,,,,,,,,,,,
```
Comment (Yuri05) [inline examples/advanced_01/input/DDI_Ratio_CYP3A4.csv]: Finally, observed data format for DDI ratio plots will look like this.
Particularly, ID is an integer now.

File: examples/advanced_01/input/master_quali_plan.json [hunk 2]
```diff
@@ -77,6 +78,104 @@
         "WatermarkSize": 40
       }
     },
+    "AxesSettings": {
```
Comment (Yuri05) [inline examples/advanced_01/input/master_quali_plan.json]: Global plot axes settings

File: examples/advanced_01/input/master_quali_plan.json [hunk 3]
```diff
@@ -867,7 +966,7 @@
               {
               "Output": "PeripheralVenousBlood|Plasma|Midazolam",
               "ObservedData": "PK_Ratio_CYP3A4",
-              "ObservedDataId": "Olkkola 1996 (Day 4)",
+              "ObservedDataId": 378,
```
Comment (Yuri05) [inline examples/advanced_01/input/master_quali_plan.json]: Id of observed data recors within observed data set is an integer now

File: schemas/OSP_Qualification_Plan_Schema.json [hunk 4]
```diff
@@ -57,6 +57,157 @@
...
+    "fractionUnit": {
+      "type": "string",
+      "enum": ["", "%"]
+    },
+    "massUnit": {
+      "type": "string",
+      "enum": ["kg", "g", "mg", "µg", "ng", "pg"]
+    },
+
+    "dimensionAndUnit": {
```
Comment (Yuri05) [inline schemas/OSP_Qualification_Plan_Schema.json]: This defines combination of properties {"Dimension", "Unit"}, which will be used in the defintion of plot axes (but could be also used elsewhere).
Currently defined for all potential dimensions which can occur on X or Y-Axis of a plot. 
Units are defined dimension-specific via `allOf+IfThenElse `
(s. https://json-schema.org/understanding-json-schema/reference/conditionals.html for details)

File: schemas/OSP_Qualification_Plan_Schema.json [hunk 5]
```diff
@@ -57,6 +57,157 @@
...
+          },
+          "then": {
+            "properties": { "Unit": { "type": "string", "enum": [""] } }
+          }
+        }
+      ],
+      "required": ["Dimension", "Unit"]
+    },
+
+    "axis": {
```
Comment (Yuri05) [inline schemas/OSP_Qualification_Plan_Schema.json]: definition of a single plot axis

File: schemas/OSP_Qualification_Plan_Schema.json [hunk 6]
```diff
@@ -57,6 +57,157 @@
...
+              "type": "string",
+              "enum": ["Linear", "Log"]
+            }
+          }
+        }
+      ],
+      "required": ["GridLines", "Scaling" ]
+    },
+
+    "axesXY": {
```
Comment (Yuri05) [inline schemas/OSP_Qualification_Plan_Schema.json]: definition of {X_axis, Y_axis} - pair

File: schemas/OSP_Qualification_Plan_Schema.json [hunk 7]
```diff
@@ -155,6 +306,19 @@
           "required": ["ChartWidth", "ChartHeight", "Fonts"]
         },
 
+        "AxesSettings": {
```
Comment (Yuri05) [inline schemas/OSP_Qualification_Plan_Schema.json]: Global axes settings.
Currently defined as mandatory (for each plot, either global or local axes settings must be available)

File: schemas/OSP_Qualification_Plan_Schema.json [hunk 8]
```diff
@@ -181,6 +345,8 @@
                   "type": "string",
                   "enum": ["predictedVsObserved", "residualsOverTime", "predictedVsObserved|residualsOverTime"]
                 },
+                "AxesPredictedVsObserved": { "$ref": "#/definitions/axesXY" },
+                "AxesResidualsOverTime": { "$ref": "#/definitions/axesXY" },
```
Comment (Yuri05) [inline schemas/OSP_Qualification_Plan_Schema.json]: OPTIONAL local axes settings per plot (will override global axes settings if given)

File: schemas/OSP_Qualification_Plan_Schema.json [hunk 9]
```diff
@@ -311,9 +480,44 @@
               },
               "required": ["SectionId", "Caption", "PKParameter", "PlotType", "Groups"]
             }
+        },
+
+        "PKRatioPlots": {
```
Comment (Yuri05) [inline schemas/OSP_Qualification_Plan_Schema.json]: Last missing plot definition: PKRatio

File: examples/advanced_01/input/master_quali_plan.json [hunk 10]
```diff
@@ -77,6 +78,104 @@
...
+        },
+        {
+          "Unit": "µg/ml",
+          "Dimension": "Concentration (mass)",
+          "Type": "Y",
+          "GridLines": false,
+          "Scaling": "Log"
+        }
+      ],
+      "DDIRatioPlots_PredictedVsObserved": [
```
Comment (msevestre) [inline examples/advanced_01/input/master_quali_plan.json]: Please do not use _ in naming of Nodes.

File: schemas/OSP_Qualification_Plan_Schema.json [hunk 11]
```diff
@@ -155,6 +306,19 @@
           "required": ["ChartWidth", "ChartHeight", "Fonts"]
         },
 
+        "AxesSettings": {
+          "type": "object",
+          "properties": {
+            "GOFMergedPlots_PredictedVsObserved": { "$ref": "#/definitions/axesXY" },
```
Comment (msevestre) [inline schemas/OSP_Qualification_Plan_Schema.json]: Please do not use _ in naming of Nodes.

File: schemas/OSP_Qualification_Plan_Schema.json [hunk 12]
```diff
@@ -277,7 +446,7 @@
                             {
                               "Output": { "$ref": "#/definitions/nonEmptyString" },
                               "ObservedData": { "$ref": "#/definitions/nonEmptyString" },
-                              "ObservedDataId": { "$ref": "#/definitions/nonEmptyString" },
+                              "ObservedDataId": { "type": "integer" },
```
Comment (msevestre) [inline schemas/OSP_Qualification_Plan_Schema.json]: What's this id here? Very unclear considering that ObservedData is already an Id

Comment (Yuri05) [reply schemas/OSP_Qualification_Plan_Schema.json]: _ObservedData_ refers to a table, _ObservedDataId_ to a record withing this table.
Could be also called _ObservedData**Record**Id_ or something like that.

File: schemas/OSP_Qualification_Plan_Schema.json (hunk 11 above)
Comment (Yuri05) [reply schemas/OSP_Qualification_Plan_Schema.json]: Don't you think it's better with underscore?
`GOFMergedPlotsPredictedVsObserved `is difficult to read

Comment (msevestre) [reply schemas/OSP_Qualification_Plan_Schema.json]: Using a consistent nomenclature allows for discoverability and is predictable

If we do not follow a convention, then more inconsistency will follow

GOF_MergedPlots_PredictedVsObserved
GOFMerged_Plots_PredictedVsObserved
GOFMergedPlots_Predicted_Vs_Observed
etc.

so in my opinion we should stick with PascalCase throughout the document

Comment (Yuri05) [reply schemas/OSP_Qualification_Plan_Schema.json]: Renamed

File: schemas/OSP_Qualification_Plan_Schema.json (hunk 12 above)
Comment (msevestre) [reply schemas/OSP_Qualification_Plan_Schema.json]: YEah I like the RecordId better. I thought it was a typo at the beginning

Comment (msevestre) [reply schemas/OSP_Qualification_Plan_Schema.json]: @Yuri05 do you want to change that in the current PR or crate another one?

Comment (Yuri05) [reply schemas/OSP_Qualification_Plan_Schema.json]: Another one. Please create an issue for this

Comment (msevestre) [reply schemas/OSP_Qualification_Plan_Schema.json]: https://github.com/Open-Systems-Pharmacology/QualificationPlan/issues/91

**Pull request merged**